processor/src/test/java/com/linecorp/decaton/processor/processors/BatchingProcessorTest.java

```diff
@@ -0,0 +1,117 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor.processors;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.verify;
+import static org.mockito.internal.verification.VerificationModeFactory.times;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+import com.linecorp.decaton.processor.Completion;
+import com.linecorp.decaton.processor.ProcessingContext;
+import com.linecorp.decaton.protocol.Sample.HelloTask;
+
+public class BatchingProcessorTest {
+    @Rule
+    public MockitoRule rule = MockitoJUnit.rule();
+
+    @Mock
+    private ProcessingContext<HelloTask> context;
+
+    @Mock
+    private Completion completion;
+
+    private final List<HelloTask> processedTasks = new ArrayList<>();
+
+    @Before
+    public void before() {
+        doReturn(completion).when(context).deferCompletion();
+        processedTasks.clear();
+    }
+
+    private HelloTask buildHelloTask(String name, int age) {
+        return HelloTask.newBuilder().setName(name).setAge(age).build();
+    }
+
+    private BatchingProcessor<HelloTask> buildProcessor(long lingerMs, int capacity) {
+        return new BatchingProcessor<HelloTask>(lingerMs, capacity) {
+            @Override
+            void processBatchingTasks(List<BatchingTask<HelloTask>> batchingTasks) {
+                batchingTasks.forEach(batchingTask -> {
+                    processedTasks.add(batchingTask.task);
+                    batchingTask.completion.complete();
+                });
+            }
+        };
+    }
+
+    @Test(timeout = 5000)
+    public void testLingerLimit() throws InterruptedException {
+        long lingerMs = 500;
+        BatchingProcessor<HelloTask> processor = buildProcessor(lingerMs, Integer.MAX_VALUE);
+
+        HelloTask task1 = buildHelloTask("one", 1);
+        HelloTask task2 = buildHelloTask("two", 2);
+        HelloTask task3 = buildHelloTask("three", 3);
+        HelloTask task4 = buildHelloTask("four", 4);
+
+        processor.process(context, task1);
+        processor.process(context, task2);
+        Thread.sleep(lingerMs * 2); // doubling just to make sure flush completed in background
```

Comment:
Using Thread.sleep to wait something happen makes test result flaky.
Let's use CountDownLatch to make sure the condition happens. (Referring CompactionProcessorTest may be helpful)

Reply:
Thank you for your reference.
I fixed it so that the test result is stable as much as possible.
357f2fb